tools/functional-tester/etcd-tester/lease_stresser.go

```diff
@@ -0,0 +1,281 @@
+		return false, err
+	}
+	return len(resp.Kvs) == 0, nil
+}
+
+func (ls *leaseStresser) createLeaseAndKeepAlive(ctx context.Context) (int64, error) {
+	leaseID, err := ls.createLease(ctx)
+	if err != nil {
+		return -1, err
+	}
+	// keep track of all the keep lease alive go routines
+	ls.aliveWg.Add(1)
+	go ls.keepLeaseAlive(ctx, leaseID)
+	return leaseID, nil
+}
+
+func (ls *leaseStresser) createLease(ctx context.Context) (int64, error) {
+	resp, err := ls.lc.LeaseGrant(ctx, &pb.LeaseGrantRequest{
+		TTL: TTL,
+	})
+	if err != nil {
+		return -1, err
+	}
+	return resp.ID, nil
+}
+
+func (ls *leaseStresser) keepLeaseAlive(ctx context.Context, leaseID int64) {
+	defer ls.aliveWg.Done()
+	for {
+		select {
+		case <-time.After(1 * time.Second):
+		case <-ctx.Done():
+			return
+		}
+		stream, err := ls.lc.LeaseKeepAlive(ctx, grpc.FailFast(false))
+		if err != nil {
+			plog.Error("keepLeaseAlive lease %v creates stream error: %v", leaseID, err)
+			return
+		}
+		err = stream.Send(&pb.LeaseKeepAliveRequest{ID: leaseID})
+		if err != nil {
+			plog.Error("keepLeaseAlive stream sends lease %v error %v ", leaseID, err)
+			continue
+		}
+		respRC, err := stream.Recv()
+		if err != nil {
+			plog.Error("keepLeaseAlive stream receives lease %v stream error %v ", leaseID, err)
+			continue
+		}
+		// lease expires after TTL become 0
+		// don't send keepalive if the lease has expired
+		if respRC.TTL <= 0 {
+			plog.Infof("keepLeaseAlive stream receives lease %v has TTL <= 0 ", leaseID)
+			return
+		}
+	}
+}
+
+// attachKeysWithLease function attaches keys to the lease.
+// the format of key is the concat of leaseID + '_' + '<order of key creation>'
+// e.g 5186835655248304152_0 for first created key and 5186835655248304152_1 for second created key
+func (ls *leaseStresser) attachKeysWithLease(ctx context.Context, leaseID int64) {
+	for j := 0; j < ls.keysPerLease; j++ {
+		// keep retrying until put succeeded for sure assuming etcd servers will be recovered eventually
+		for {
+			_, err := ls.kvc.Put(ctx, &pb.PutRequest{
+				Key:   []byte(fmt.Sprintf("%d%s%d", leaseID, "_", j)),
+				Value: []byte(fmt.Sprintf("bar")),
+				Lease: leaseID,
+			})
+			if err == nil {
+				return
+			}
+		}
+	}
+}
+
+// randomlyDropLease drops the lease only when the rand.Int(2) returns 1.
+// This creates a 50/50 percents chance of dropping a lease
+func (ls *leaseStresser) randomlyDropLease(ctx context.Context, leaseID int64) bool {
+	if rand.Intn(2) != 0 {
+		return false
+	}
+	// keep retrying until a lease is dropped for sure assuming the etcd servers will eventually recover
+	for {
+		resp, err := ls.lc.LeaseRevoke(ctx, &pb.LeaseRevokeRequest{ID: leaseID})
+		plog.Errorf("revoking lease %v resp %v error %v ", leaseID, resp, err)
+		if err == nil || rpctypes.Error(err) == rpctypes.ErrLeaseNotFound {
+			return true
+		}
+	}
+}
+
+func (ls *leaseStresser) Cancel() {
+	ls.cancel()
+	ls.conn.Close()
+	ls.aliveWg.Wait()
```

Comment:
I decided not implement Cancel() fully at the moment. because the leaseStresser does not cancel after each round.